### ISSUE TITLE
Download ECNF, sage issue

### ISSUE THREAD

**Issue by edgarcosta** [LMFDB/lmfdb#4323]
About once a day, a bot manages to trigger a weird sage bug (traceback below), which I can't replicate.
Does anyone know the root of this issue? It would be nice to report this upstream.

 
```
Exception on /EllipticCurve/6.6.371293.1/27.1/a/2/download/gp [GET]
Traceback (most recent call last):
  File "sage/structure/category_object.pyx", line 839, in sage.structure.category_object.CategoryObject.getattr_from_category (build/cythonized/sage/structure/category_object.c:7075)
    return self.__cached_methods[name]
KeyError: '_pari_bnf'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/sage/rings/number_field/number_field.py", line 4239, in pari_bnf
    bnf = self._pari_bnf
  File "sage/structure/category_object.pyx", line 833, in sage.structure.category_object.CategoryObject.__getattr__ (build/cythonized/sage/structure/category_object.c:6994)
    return self.getattr_from_category(name)
  File "sage/structure/category_object.pyx", line 848, in sage.structure.category_object.CategoryObject.getattr_from_category (build/cythonized/sage/structure/category_object.c:7160)
    attr = getattr_from_other_class(self, cls, name)
  File "sage/cpython/getattr.pyx", line 367, in sage.cpython.getattr.getattr_from_other_class (build/cythonized/sage/cpython/getattr.c:2553)
    raise AttributeError(dummy_error_message)
AttributeError: 'NumberField_absolute_with_category' object has no attribute '_pari_bnf'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/lmfdb/lmfdb-git-web/lmfdb/ecnf/main.py", line 676, in ecnf_code_download
    response = make_response(ecnf_code(**args))
  File "/home/lmfdb/lmfdb-git-web/lmfdb/ecnf/main.py", line 712, in ecnf_code
    E = ECNF.by_label(label)
  File "/home/lmfdb/lmfdb-git-web/lmfdb/ecnf/WebEllipticCurve.py", line 262, in by_label
    return ECNF(data)
  File "/home/lmfdb/lmfdb-git-web/lmfdb/ecnf/WebEllipticCurve.py", line 253, in __init__
    self.make_E()
  File "/home/lmfdb/lmfdb-git-web/lmfdb/ecnf/WebEllipticCurve.py", line 282, in make_E
    self.cond = pretty_ideal(N)
  File "/home/lmfdb/lmfdb-git-web/lmfdb/ecnf/WebEllipticCurve.py", line 111, in pretty_ideal
    gens = I.gens_reduced() if easy else I.gens()
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/sage/rings/number_field/number_field_ideal.py", line 793, in gens_reduced
    self._cache_bnfisprincipal(proof=proof, gens_needed=True)
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/sage/rings/number_field/number_field_ideal.py", line 1063, in _cache_bnfisprincipal
    bnf = self.number_field().pari_bnf(proof)
  File "/home/sage/sage-9.2/local/lib/python3.8/site-packages/sage/rings/number_field/number_field.py", line 4243, in pari_bnf
    self._pari_bnf = f.bnfinit(1)
  File "cypari2/auto_gen.pxi", line 3479, in cypari2.gen.Gen_base.bnfinit
RuntimeError: Aborted
[2021-01-06 00:54:36 UTC] 500 error on URL https://www.lmfdb.org/EllipticCurve/6.6.371293.1/27.1/a/2/download/gp ()
```

**Comment by JohnCremona** [comment]
I'll take a look. It is certainly not great that the web elliptic curve constructor for ECNF triggers a call to obtain generators of an ideal: if we want to display them we should store them. Mea culpa

**Comment by AndrewVSutherland** [comment]
It would be great to find away to avoid these errors, even if the problem is in Sage or pari -- as Edgar noted, we get emails about these errors much every day or two, which makes it easy for us to miss non-empty flasklog alerts that are more serious.

**Comment by JohnCremona** [comment]
It is certainly my fault that we see this error now:   lines 110-111 of WebEllipticCurve.py show that it used to only compute "reduced_gens" (in particular, this gives a single generator for principal ideals) over quadratic fields or when the ideal has norm 1.  Now it always tries (but not in a try-except block).   Note that the db does have stored the conductor's gens, but the problematic function is called also for 3 other ideals (a non-minimal prime when there is one, the model discriminant ideal, and the minimial discriminant ideal -- for these, the generators are not stored.

When I run this myself (on legendre) the curve's home page displays fine.  Another question is: when the request is for the /download/gp and not the home page, why is the home page itself being created?

**Comment by JohnCremona** [comment]
As it would be a fair amount of work to make this class not (ever) create the base number field or ideals in it, perhaps as a temporary measure I could reinstate the test that we do not call this ideal_pretty() function when the base field has degree greater than 2 (or 3 or 4 are probably harmless).  Note that nothing I have said here explains why an error is ever raised!

**Comment by AndrewVSutherland** [comment]
As Edgar said, we can't seem to reproduce the problem (when I visit the link in the flasklog it always works fine).  I'm not sure why it is creating the page, I don't think this happens for download links in other sections.

**Comment by JohnCremona** [comment]
The function ecnf_code_download() calls ecnf_code() which does create the class instance, as a cheap (to code) way of getting the code snippets, but that is wasteful for anyone who does not get to this .../download/gp path by following a link on the curve's home page.  I don't think humans would do this, but I can imagine someone writing a pari/gp function to pull a curve (/NF) from the website using this.

I can (and will) change the implementation of ecnf_code to not create the ENNF object.

**Comment by edgarcosta** [comment]
Someone managed to trigger the same issue on https://beta.lmfdb.org/EllipticCurve/6.6.371293.1/27.1/a/2
```
Exception on /EllipticCurve/6.6.371293.1/27.1/a/2 [GET]
Traceback (most recent call last):
  File "sage/structure/category_object.pyx", line 839, in sage.structure.category_object.CategoryObject.getattr_from_category (build/cythonized/sage/structure/category_object.c:7075)
    return self.__cached_methods[name]
KeyError: '_pari_bnf'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field.py", line 4239, in pari_bnf
    bnf = self._pari_bnf
  File "sage/structure/category_object.pyx", line 833, in sage.structure.category_object.CategoryObject.__getattr__ (build/cythonized/sage/structure/category_object.c:6994)
    return self.getattr_from_category(name)
  File "sage/structure/category_object.pyx", line 848, in sage.structure.category_object.CategoryObject.getattr_from_category (build/cythonized/sage/structure/category_object.c:7160)
    attr = getattr_from_other_class(self, cls, name)
  File "sage/cpython/getattr.pyx", line 367, in sage.cpython.getattr.getattr_from_other_class (build/cythonized/sage/cpython/getattr.c:2553)
    raise AttributeError(dummy_error_message)
AttributeError: 'NumberField_absolute_with_category' object has no attribute '_pari_bnf'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/main.py", line 359, in show_ecnf
    ec = ECNF.by_label(label)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 262, in by_label
    return ECNF(data)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 253, in __init__
    self.make_E()
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 282, in make_E
    self.cond = pretty_ideal(N)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 111, in pretty_ideal
    gens = I.gens_reduced() if easy else I.gens()
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field_ideal.py", line 793, in gens_reduced
    self._cache_bnfisprincipal(proof=proof, gens_needed=True)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field_ideal.py", line 1063, in _cache_bnfisprincipal
    bnf = self.number_field().pari_bnf(proof)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field.py", line 4243, in pari_bnf
    self._pari_bnf = f.bnfinit(1)
  File "cypari2/auto_gen.pxi", line 3479, in cypari2.gen.Gen_base.bnfinit
RuntimeError: Aborted
[2021-01-06 14:46:01 UTC] 500 error on URL https://beta.lmfdb.org/EllipticCurve/6.6.371293.1/27.1/a/2 ()
Exception on /EllipticCurve/6.6.371293.1/27.1/a/2 [GET]
Traceback (most recent call last):
  File "sage/structure/category_object.pyx", line 839, in sage.structure.category_object.CategoryObject.getattr_from_category (build/cythonized/sage/structure/category_object.c:7075)
    return self.__cached_methods[name]
KeyError: '_pari_bnf'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field.py", line 4239, in pari_bnf
    bnf = self._pari_bnf
  File "sage/structure/category_object.pyx", line 833, in sage.structure.category_object.CategoryObject.__getattr__ (build/cythonized/sage/structure/category_object.c:6994)
    return self.getattr_from_category(name)
  File "sage/structure/category_object.pyx", line 848, in sage.structure.category_object.CategoryObject.getattr_from_category (build/cythonized/sage/structure/category_object.c:7160)
    attr = getattr_from_other_class(self, cls, name)
  File "sage/cpython/getattr.pyx", line 367, in sage.cpython.getattr.getattr_from_other_class (build/cythonized/sage/cpython/getattr.c:2553)
    raise AttributeError(dummy_error_message)
AttributeError: 'NumberField_absolute_with_category' object has no attribute '_pari_bnf'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/main.py", line 359, in show_ecnf
    ec = ECNF.by_label(label)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 262, in by_label
    return ECNF(data)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 253, in __init__
    self.make_E()
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 282, in make_E
    self.cond = pretty_ideal(N)
  File "/home/lmfdb/lmfdb-git-dev/lmfdb/ecnf/WebEllipticCurve.py", line 111, in pretty_ideal
    gens = I.gens_reduced() if easy else I.gens()
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field_ideal.py", line 793, in gens_reduced
    self._cache_bnfisprincipal(proof=proof, gens_needed=True)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field_ideal.py", line 1063, in _cache_bnfisprincipal
    bnf = self.number_field().pari_bnf(proof)
  File "/home/sage/sage-9.2/local/lib/python3.6/site-packages/sage/rings/number_field/number_field.py", line 4243, in pari_bnf
    self._pari_bnf = f.bnfinit(1)
  File "cypari2/auto_gen.pxi", line 3479, in cypari2.gen.Gen_base.bnfinit
RuntimeError: Aborted
[2021-01-06 15:39:19 UTC] 500 error on URL https://beta.lmfdb.org/EllipticCurve/6.6.371293.1/27.1/a/2 ()
```

**Comment by JohnCremona** [comment]
That would have been me....

I am working on this, will not be much longer!

**Comment by JohnCremona** [comment]
It is taking longer than expected since I discovered that after some schema changes a couple of months ago the code converting code snippet templates into working code was not working.  I may not finish this today.

**Comment by JohnCremona** [comment]
I wrote "I may not finish this today" 22 days ago, and I was right.

I have just uploaded new data for all ECNF curves.  Here are the changes:
   - there are 2 new columns, 'disc' (holds the discriminant ideal of the model), 'normdisc' (its norm)
   - all ideals stored in the database are stored as strings (gen) or (gen1,gen2), the first iff it is principle, which tidies up displays and also avoids any ideal computation on the fly.  For principle ideals, the generator is "reducied modulo units" in a well-defined way, and in particular the unit ideal is '(1)'.  The ideals in question are: conductor, model discriminant, minimal discriminant, each bad prime.
   - Kodaira symbols are now stored as ints (pari encoding) as for EC/Q.

No code changes are needed (after #4324, #4329 which were backwards and forwards compatible), but some code simplification is now possible, which I will get to at some point.   Otherwise I would like to leave the updated data as it is on beta for a few days before copying ec_nfcurves to prod, just in case.
